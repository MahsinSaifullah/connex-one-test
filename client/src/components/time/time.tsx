import React from 'react';

export const Time: React.FC = () => {
  return <div>time</div>;
};
