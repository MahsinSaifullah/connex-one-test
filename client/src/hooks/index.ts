export * from './time.hook';
export * from './metric.hook';
