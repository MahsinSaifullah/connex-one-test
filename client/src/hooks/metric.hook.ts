import * as React from 'react';
import { getMetrics } from '../services';

export const useMetric = () => {
  const [metrics, setMetrics] = React.useState<string>('');
  const [isLoading, setIsLoading] = React.useState(false);
  const [isError, setIsError] = React.useState(false);
  const metricFetchRef = React.useRef<NodeJS.Timer>();

  const getPromMetrics = async () => {
    try {
      setIsLoading(true);
      const _metrics = await getMetrics();
      setMetrics(_metrics);
    } catch (error) {
      setIsError(true);
    }
    setIsLoading(false);
  };

  React.useEffect(() => {
    getPromMetrics();

    if (metricFetchRef.current) {
      clearInterval(metricFetchRef.current);
    }

    metricFetchRef.current = setInterval(async () => {
      setIsError(false);
      await getPromMetrics();
    }, 30000);

    return () => {
      clearInterval(metricFetchRef.current);
    };
  }, []);

  return {
    metrics,
    isError,
    isLoading,
  };
};
