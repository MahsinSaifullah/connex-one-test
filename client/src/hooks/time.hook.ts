import * as React from 'react';

import { IServerTime, ITimeDifference } from '../models';
import { getServerTime } from '../services';
import { calculateTimeDifference, convertSecondToHourMinSec } from '../utils';

export const useTime = () => {
  const [serverTime, setServerTime] = React.useState<IServerTime>({
    epoch: 0,
  });
  const [timeDiff, setTimeDiff] = React.useState<ITimeDifference>({
    hours: 0,
    minutes: 0,
    seconds: 0,
  });
  const [isLoading, setIsLoading] = React.useState(false);
  const [isError, setIsError] = React.useState(false);
  const timeFetchRef = React.useRef<NodeJS.Timer>();
  const timeDiffRef = React.useRef<NodeJS.Timer>();

  const getTimeDiff = () => {
    const timeDiffFromServer = calculateTimeDifference(serverTime.epoch);
    setTimeDiff(convertSecondToHourMinSec(timeDiffFromServer));
  };

  const getTime = async () => {
    try {
      setIsLoading(true);
      const time = await getServerTime();
      setServerTime(time);
    } catch (error) {
      setIsError(true);
    }

    setIsLoading(false);
  };

  React.useEffect(() => {
    getTime();

    if (timeFetchRef.current) {
      clearInterval(timeFetchRef.current);
    }

    timeFetchRef.current = setInterval(async () => {
      await getTime();
    }, 30000);

    return () => {
      clearInterval(timeFetchRef.current);
    };
  }, []);

  React.useEffect(() => {
    if (timeDiffRef.current) {
      clearInterval(timeDiffRef.current);
    }

    timeDiffRef.current = setInterval(async () => {
      getTimeDiff();
    }, 1000);

    return () => {
      clearInterval(timeDiffRef.current);
    };
  }, [serverTime.epoch]);

  return {
    serverTime,
    timeDiff,
    isLoading,
    isError,
  };
};
