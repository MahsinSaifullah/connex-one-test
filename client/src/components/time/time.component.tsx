import * as React from 'react';

import { useTime } from '../../hooks';
import { Card, Loading, Title } from '../shared';

export const Time: React.FC = () => {
  const { timeDifferenceDisplay, serverTime, isLoading } = useTime();

  return (
    <div className="flex flex-col justify-center items-center space-y-4">
      <Title title="Server Timestamp" />
      <Card className="sm:min-w-[600px] h-[400px] sm:h-[500px] max-w-sm sm:max-w-2xl flex flex-col space-y-10 p-10 justify-center items-center">
        {isLoading ? (
          <Loading loadingText="Updating" />
        ) : (
          <>
            <div className="text-sm sm:text-lg tracking-wider">
              <span className="text-sm sm:text-2xl bg-black text-white mr-2 p-2 rounded">
                Server Timestamp:
              </span>{' '}
              {serverTime.epoch}
            </div>
            <div className="text:sm sm:text-lg tracking-wider">
              <span className="text-sm sm:text-2xl bg-rose-700 text-white mr-2 p-2 rounded">
                Time Difference:
              </span>{' '}
              {timeDifferenceDisplay}
            </div>
          </>
        )}
      </Card>
    </div>
  );
};
