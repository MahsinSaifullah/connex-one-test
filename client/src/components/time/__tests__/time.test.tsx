import { render } from '@testing-library/react';
import { screen } from '@testing-library/dom';

import * as hooks from '../../../hooks';
import { Time } from '../time.component';

jest.mock('../../../hooks', () => {
  return {
    __esModule: true,
    ...jest.requireActual('../../../hooks'),
  };
});

const spiedTimeHook = jest.spyOn(hooks, 'useTime');

describe('Time Component', () => {
  it('should render loading component if isLoading is true', () => {
    spiedTimeHook.mockReturnValue({
      serverTime: { epoch: 0 },
      timeDifferenceDisplay: '',
      isLoading: true,
      isError: false,
    });

    const { asFragment } = render(<Time />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render time component if isLoading is false', () => {
    spiedTimeHook.mockReturnValue({
      serverTime: { epoch: 100 },
      timeDifferenceDisplay: '0:0:0',
      isLoading: false,
      isError: false,
    });

    const { asFragment } = render(<Time />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correct server time and time difference', () => {
    spiedTimeHook.mockReturnValue({
      serverTime: { epoch: 100 },
      timeDifferenceDisplay: '10:10:10',
      isLoading: false,
      isError: false,
    });

    render(<Time />);

    const serverTimeElement = screen.getByTestId('server-time');
    const timeDifferenceElement = screen.getByTestId('time-difference');

    expect(serverTimeElement.innerHTML).toBe('100');
    expect(timeDifferenceElement.innerHTML).toBe('10:10:10');
  });
});
