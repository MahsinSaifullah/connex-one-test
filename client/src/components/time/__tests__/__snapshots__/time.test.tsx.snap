// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time Component should render loading component if isLoading is true 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col justify-center items-center space-y-4"
  >
    <h2
      class="text-lg sm:text-2xl font-bold tracking-widest undefined"
    >
      Server Timestamp
    </h2>
    <article
      class="rounded shadow-lg bg-white overflow-scroll sm:min-w-[600px] h-[400px] xl:h-[500px] max-w-sm lg:max-w-2xl flex flex-col space-y-10 p-10 justify-center items-center"
    >
      <div
        class="text-2xl animate-pulse undefined"
      >
        Updating ...
      </div>
    </article>
  </div>
</DocumentFragment>
`;

exports[`Time Component should render time component if isLoading is false 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col justify-center items-center space-y-4"
  >
    <h2
      class="text-lg sm:text-2xl font-bold tracking-widest undefined"
    >
      Server Timestamp
    </h2>
    <article
      class="rounded shadow-lg bg-white overflow-scroll sm:min-w-[600px] h-[400px] xl:h-[500px] max-w-sm lg:max-w-2xl flex flex-col space-y-10 p-10 justify-center items-center"
    >
      <div
        class="text-sm sm:text-lg tracking-wider"
      >
        <span
          class="text-sm sm:text-2xl bg-black text-white mr-2 p-2 rounded"
        >
          Server Timestamp:
        </span>
         
        <span
          data-testid="server-time"
        >
          100
        </span>
      </div>
      <div
        class="text:sm sm:text-lg tracking-wider"
      >
        <span
          class="text-sm sm:text-2xl bg-rose-700 text-white mr-2 p-2 rounded"
        >
          Time Difference:
        </span>
         
        <span
          data-testid="time-difference"
        >
          0:0:0
        </span>
      </div>
    </article>
  </div>
</DocumentFragment>
`;
